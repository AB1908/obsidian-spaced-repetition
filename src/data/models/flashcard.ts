import { CardType, ReviewResponse, schedule } from "src/scheduler/scheduling";
import { nanoid } from "nanoid";
import { FLAG, FlashcardMetadata, parseCardText, parseFileText, parseMetadata } from "src/data/parser";
import { moment, Notice } from "obsidian";
import { SchedulingMetadata } from "src/data/utils/TextGenerator";
import { plugin } from "src/main";
import { ParsedCard } from "src/data/models/parsedCard";
import { getAnnotationFilePath } from "src/data/models/sourceNote";
import { filePathsWithTag, getCurrentDateIsoString } from "src/data/disk";
import { calculateDelayBeforeReview } from "src/utils";

export abstract class AbstractFlashcard {
    answerText: string;
    cardType: number;
    context: string;
    dueDate: string;
    ease: number;
    id: string;
    interval: number;
    questionText: string;
    siblings: string[];
    // todo: rename
    parentId: string;
    flag: FLAG;
    parsedCardId: string;

    // cardMetadata and highlight ID are mutually exclusive properties. Given that there is no constructor overloading
    // probably should change this to be a union type
    protected constructor(cardType: CardType, parsedCardId: string, cardMetadata?: FlashcardMetadata, annotationId?: string) {
        this.id = nanoid(8);
        this.cardType = cardType;
        this.context = null;
        this.dueDate = cardMetadata?.dueDate;
        this.ease = cardMetadata?.ease;
        this.interval = cardMetadata?.interval;
        this.parentId = cardMetadata?.annotationId || annotationId;
        this.flag = cardMetadata?.flag;
        this.siblings = [];
        this.parsedCardId = parsedCardId;
    }

    isDue() {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        return this.dueDate === null || this.dueDate <= getCurrentDateIsoString();
    }
}

export class Flashcard extends AbstractFlashcard {
    // cardMetadata and highlight ID are mutually exclusive properties. Given that there is no constructor overloading
    // probably should change this to be a union type
    constructor(parsedCardId: string, questionText: string, answerText: string, cardMetadata?: FlashcardMetadata, annotationId?: string) {
        // todo: handle other types later
        const cardType = CardType.MultiLineBasic;
        if (cardMetadata) {
            super(cardType, parsedCardId, cardMetadata);
        } else {
            super(cardType, parsedCardId, null, annotationId);
        }
        this.questionText = questionText;
        this.answerText = answerText;
    }
}

// todo: rename to update card?
export function schedulingMetadataForResponse(
    clickedResponse: ReviewResponse,
    schedulingMetadata: SchedulingMetadata,
): SchedulingMetadata {
    // const flashcard = getFlashcardById(flashcardId);
    // take the response received
    // use that to update flashcard internal state
    // that will take care of writing to disk
    // so this should be a relatively lean method
    // don't forget to update siblings?
    let schedObj;
    // is new card
    if (schedulingMetadata.dueDate === null) {
        // todo: move default settings down into schedule()?
        schedObj = schedule(
            clickedResponse,
            1.0,
            plugin.data.settings.baseEase,
            0,
        );
    } else {
        schedObj = schedule(
            clickedResponse,
            schedulingMetadata.interval,
            schedulingMetadata.ease,
            calculateDelayBeforeReview(schedulingMetadata.dueDate),
        );
    }
    const {interval, ease} = schedObj;
    // todo: parameterize format? nah
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const due = moment(Date.now() + interval * 24 * 3600 * 1000).format("YYYY-MM-DD");
    return {interval, ease, dueDate: due};
}

export function maturityCounts(flashcards: Flashcard[]) {
    let newCount = 0, mature = 0, learning = 0;
    flashcards.forEach(t => {
        if (t.interval >= 32) {
            mature += 1;
        } else if (t.interval && t.interval < 32) {
            learning += 1;
        } else {
            // } if (!t.interval) {
            newCount += 1;
        }
    });
    return {mature, learning, new: newCount};
}

export function createFlashcard(parsedCard: ParsedCard, questionText: string, answerText: string) {
    return new Flashcard(parsedCard.id, questionText, answerText, parseMetadata(parsedCard.metadataText));
}

export function generateFlashcardsArray(parsedCardsArray: ParsedCard[]) {
    const out: Flashcard[] = [];
    for (const parsedCard of parsedCardsArray) {
        const [questionText, answerText] = parseCardText(parsedCard.cardText);
        out.push(createFlashcard(parsedCard, questionText, answerText));
    }
    return out;
}

export class FlashcardNote {
    path: string;
    flashcards: Flashcard[];
    parentPath: string;
    parsedCards: ParsedCard[];

    constructor(path: string) {
        this.path = path;
        this.flashcards = [];
        this.parentPath = "";
        this.parsedCards = [];
    }

    async initialize(flashcardIndex: FlashcardIndex) {
        this.parentPath = getAnnotationFilePath(this.path);
        this.parsedCards = await parseFileText(this.path);
        try {
            this.flashcards = generateFlashcardsArray(this.parsedCards);
            this.flashcards.forEach(t => flashcardIndex.flashcards.set(t.id, t));
        } catch (e) {
            console.error(e);
            throw new Error(e);
        }
        return this;
    }
}

export class FlashcardIndex {
    flashcardNotes: FlashcardNote[];
    flashcards: Map<string, Flashcard>;

    constructor() {
        this.flashcardNotes = [];
        this.flashcards = new Map<string, Flashcard>();
    }

    async initialize() {
        // todo: remove hardcoding of tag
        const filePaths = filePathsWithTag("flashcards");
        const notesWithFlashcards = filePaths.map((t: string) => new FlashcardNote(t));
        for (const t of notesWithFlashcards) {
            try {
                await t.initialize(this);
            } catch (e) {
                // WARNING! this is dangerous, I am catching other errors and just assuming that these are this error
                console.error(e);
                console.error(`init: unable to initialize book ${t.path}`);
                new Notice("Error: Unable to parse legacy SRS flashcards. Try removing the #flashcards tag from files with SRS flashcards.");
            }
        }
        this.flashcardNotes = notesWithFlashcards;
        console.log("Card Coverage: Flashcard index successfully initialized");
        return this;
    }

    addFlashcardNoteToIndex(flashcardNote: FlashcardNote) {
        if (this.flashcardNotes.find(t => t.path === flashcardNote.path)) {
            throw new Error(`addFlashcardNoteToIndex: ${flashcardNote.path} is already in the index`);
        }
        this.flashcardNotes.push(flashcardNote);
    }
}