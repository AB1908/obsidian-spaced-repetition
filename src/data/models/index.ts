import { SourceNote } from "src/data/models/sourceNote";
import { FlashcardNote } from "src/data/models/flashcard";
import { annotation } from "src/data/models/annotations";
import { paragraph } from "src/data/models/paragraphs";

export class Index {
    sourceNoteIndex: Map<string, SourceNote>;
    flashcardNoteIndex: Map<string, FlashcardNote>;
    annotationIndex: Map<string, (paragraph|annotation)>

    constructor() {
        this.annotationIndex = new Map<string, paragraph | annotation>();
    }

    initialize() {

    }

    addToAnnotationIndex(block: annotation|paragraph) {
        // todo: think about incorporating an index dirty state to
        // properly retrigger indexing
        this.annotationIndex.set(block.id, block);
    }

    getAnnotation(blockId: string) {
        return this.annotationIndex.get(blockId);
    }
}