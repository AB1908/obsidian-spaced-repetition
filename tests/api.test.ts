import { createFlashcardForAnnotation, getFlashcardById } from "src/api";
import { plugin } from "src/main";
import { FlashcardIndex } from "src/data/models/flashcard";

jest.mock("../src/data/disk", () => {
    return {
        // eslint-disable-next-line @typescript-eslint/no-empty-function,@typescript-eslint/no-unused-vars
        writeCardToDisk: (path: string, text: string) => {
        }
    };
});

jest.mock("../src/data/models/flashcard", () => {
    // Require the original module to not be mocked...
    const originalModule = jest.requireActual<typeof import("../src/data/models/flashcard")>(
        "../src/data/models/flashcard"
    );

    return {
        __esModule: true, // Use it when dealing with esModules
        ...originalModule,
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        calculateDelayBeforeReview: jest.fn((due: string) => 63836018)
    };
});

jest.mock("../src/main", () => {
       return {
        __esModule: true, // Use it when dealing with esModules
        plugin: { flashcardIndex: null },
    };
});



describe("getFlashcardById", () => {
    beforeEach(() => {
        plugin.flashcardIndex = new FlashcardIndex();
        plugin.flashcardIndex.flashcards = new Map();
        plugin.flashcardIndex.flashcards.set("pBri5QNB",
            {
                id: "pBri5QNB",
                cardType: 2,
                context: null,
                dueDate: "2023-09-02",
                ease: 250,
                interval: 2,
                annotationId: "5769",
                flag: "L",
                siblings: [],
                parsedCardId: "u-72tWEW",
                questionText: "What is cued recall?",
                answerText:
                    "Cued recall is where we cue recall by presenting some information."
            }
        );
    });
    test("retrieves a flashcard successfully", () => {
        expect(getFlashcardById("pBri5QNB", "ibJ6QFl4")).toStrictEqual({
            id: "pBri5QNB",
            cardType: 2,
            context: null,
            dueDate: "2023-09-02",
            ease: 250,
            interval: 2,
            delayBeforeReview: 63836018,
            annotationId: "5769",
            flag: "L",
            siblings: [],
            parsedCardId: "u-72tWEW",
            questionText: "What is cued recall?",
            answerText: "Cued recall is where we cue recall by presenting some information."
        });
        expect(() => getFlashcardById("aaaa", "ibJ6QFl4")).toThrowError();
    });
});

// describe("updateFlashcardQuestion", () => {
//     let mockThis: any;
//     let boundUpdate: any;
//
//     beforeEach(() => {
//         mockThis = {
//             flashcards: flashcards()
//         };
//         boundUpdate = updateFlashcardQuestion.bind(mockThis);
//     });
//
//     test("should update the questionText of the flashcard with the given id", () => {
//         const updatedQuestion = "What is your age?";
//         const id = "yjlML2s9W";
//
//         expect(boundUpdate(id, updatedQuestion)).toBe(true);
//         expect(mockThis.flashcards[0].questionText).toBe(updatedQuestion);
//     });
//
//     test("should return false if the flashcard with the given id does not exist", () => {
//         const nonExistingId = "3";
//         const updatedQuestion = "What is your age?";
//
//         expect(boundUpdate(nonExistingId, updatedQuestion)).toBe(false);
//     });
// });
//
// describe("updateFlashcardAnswer", () => {
//     let mockThis: any;
//     let boundUpdate: any;
//
//     beforeEach(() => {
//         mockThis = {
//             flashcards: flashcards()
//         };
//         boundUpdate = updateFlashcardAnswer.bind(mockThis);
//     });
//
//     test("should update the questionText of the flashcard with the given id", () => {
//         const updatedAnswer = "What is your age?";
//         const id = "yjlML2s9W";
//
//         expect(boundUpdate(id, updatedAnswer)).toBe(true);
//         expect(mockThis.flashcards[0].answerText).toBe(updatedAnswer);
//     });
//
//     test("should return false if the flashcard with the given id does not exist", () => {
//         const nonExistingId = "3";
//         const updatedQuestion = "What is your age?";
//
//         expect(boundUpdate(nonExistingId, updatedQuestion)).toBe(false);
//     });
// });

describe("createFlashcardForAnnotation", () => {
    test("should create a new flashcard", async () => {
        const card = {
            id: "pBri5QNB",
            cardType: 2,
            context: null,
            dueDate: "2023-09-02",
            ease: 250,
            interval: 2,
            delayBeforeReview: 63836018,
            annotationId: "5769",
            flag: "L",
            siblings: [],
            parsedCardId: "u-72tWEW",
            questionText: "What is cued recall?",
            answerText: "Cued recall is where we cue recall by presenting some information."
        };
        const bookId = "ibJ6QFl4";
        await createFlashcardForAnnotation(
            card.questionText,
            card.answerText,
            card.annotationId,
            bookId
        );
        // console.log(plugin.notesWithFlashcards);
        expect(plugin.notesWithFlashcards).toMatchInlineSnapshot(`
            [
              {
                "flashcards": [
                  {
                    "annotationId": "5769",
                    "answerText": "Cued recall is where we cue recall by presenting some information.",
                    "cardType": 2,
                    "context": null,
                    "dueDate": "2023-09-02",
                    "ease": 250,
                    "flag": "L",
                    "id": "pBri5QNB",
                    "interval": 2,
                    "parsedCardId": "u-72tWEW",
                    "questionText": "What is cued recall?",
                    "siblings": [],
                  },
                  Flashcard {
                    "annotationId": "5769",
                    "answerText": "Cued recall is where we cue recall by presenting some information.",
                    "cardType": 2,
                    "context": null,
                    "dueDate": undefined,
                    "ease": undefined,
                    "flag": undefined,
                    "id": "1",
                    "interval": undefined,
                    "parsedCardId": "0",
                    "questionText": "What is cued recall?",
                    "siblings": [],
                  },
                ],
                "flashcardsPath": "Memory - A Very Short Introduction/Flashcards.md",
                "id": "ibJ6QFl4",
                "name": "Memory - A Very Short Introduction",
                "parsedCards": [
                  {
                    "cardText": "What is cued recall?
            ?
            Cued recall is where we cue recall by presenting some information.",
                    "cardType": 2,
                    "id": "u-72tWEW",
                    "lineNo": -1,
                    "metadataText": "<!--SR:5769!L,2023-09-02,2,250-->",
                    "notePath": "Memory - A Very Short Introduction/Flashcards.md",
                  },
                  {
                    "cardText": "What is cued recall?
            ?
            Cued recall is where we cue recall by presenting some information.",
                    "cardType": 2,
                    "id": "0",
                    "lineNo": -1,
                    "metadataText": "<!--SR:5769-->",
                    "notePath": "Memory - A Very Short Introduction/Flashcards.md",
                  },
                ],
              },
            ]
        `);
    });
});

// describe("deleteFlashcard", () => {
//     let mockThis: any;
//     let boundDelete: typeof deleteFlashcardById;
//
//     beforeEach(() => {
//         mockThis = {
//             flashcards: flashcards()
//         };
//         boundDelete = deleteFlashcardById.bind(mockThis);
//     });
//
//     test("should delete an existing flashcard", () => {
//         const id = "yjlML2s9W";
//         expect(boundDelete(id)).toBe(true);
//         expect(mockThis.flashcards).toStrictEqual([]);
//     });
//
//     test("should return false if flashcard doesn't exist", () => {
//         const id = "aaaa";
//         expect(boundDelete(id)).toBe(false);
//         expect(mockThis.flashcards).toStrictEqual(flashcards());
//     });
//
//     test("should throw an error when trying to delete from an empty flashcard array", () => {
//         mockThis.flashcards = [];
//         boundDelete = deleteFlashcardById.bind(mockThis);
//         const id = "aaaa";
//         expect(() => boundDelete(id)).toThrow();
//     });
// });
