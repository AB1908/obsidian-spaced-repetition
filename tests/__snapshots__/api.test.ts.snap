// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSectionTreeForBook generates correct JSON for book tree 1`] = `
{
  "children": [
    {
      "children": [],
      "counts": {
        "with": 2,
        "without": 7,
      },
      "id": "93",
      "level": 1,
      "name": "Chapter 1: You are your memory",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 1,
          },
          "id": "95",
          "level": 2,
          "name": "The logic of memory: encoding, storage, and retrieval",
        },
        {
          "children": [],
          "counts": {
            "with": 7,
            "without": 23,
          },
          "id": "96",
          "level": 2,
          "name": "Different kinds of memory: the functional structure of remembering",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 7,
          },
          "id": "97",
          "level": 2,
          "name": "Semantic, episodic, and procedural memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "98",
          "level": 2,
          "name": "Explicit and implicit memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 4,
          },
          "id": "99",
          "level": 2,
          "name": "Different kinds of memory task:",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "100",
          "level": 2,
          "name": "The experience of memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "101",
          "level": 2,
          "name": "Levels of processing",
        },
      ],
      "counts": {
        "with": 9,
        "without": 41,
      },
      "id": "94",
      "level": 1,
      "name": "Chapter 2: Mapping memories",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 0,
          },
          "id": "103",
          "level": 2,
          "name": "Inferring memory from behaviour",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "105",
              "level": 3,
              "name": "Context retrieval",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "106",
              "level": 3,
              "name": "Familiarity",
            },
          ],
          "counts": {
            "with": 3,
            "without": 6,
          },
          "id": "104",
          "level": 2,
          "name": "Retrieval: recall versus recognition",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 4,
          },
          "id": "107",
          "level": 2,
          "name": "The effect of context on recall and recognition",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "108",
          "level": 2,
          "name": "Unconscious influences on memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 6,
          },
          "id": "109",
          "level": 2,
          "name": "Categories versus continuum?",
        },
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 1,
          },
          "id": "110",
          "level": 2,
          "name": "Relating study and test",
        },
      ],
      "counts": {
        "with": 6,
        "without": 19,
      },
      "id": "102",
      "level": 1,
      "name": "Chapter 3: Pulling the rabbit out of the hat",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 3,
            "without": 3,
          },
          "id": "112",
          "level": 2,
          "name": "Forgetting",
        },
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 4,
          },
          "id": "113",
          "level": 2,
          "name": "Flashbulb memories and the reminiscence bump",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 4,
          },
          "id": "114",
          "level": 2,
          "name": "Organization and errors in memory",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 15,
              },
              "id": "116",
              "level": 3,
              "name": "Schemas – what we already know:",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "117",
              "level": 3,
              "name": "How does knowledge promote remembering?",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 4,
              },
              "id": "118",
              "level": 3,
              "name": "How can knowledge lead to errors?",
            },
          ],
          "counts": {
            "with": 0,
            "without": 21,
          },
          "id": "115",
          "level": 2,
          "name": "The effects of previous knowledge",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 5,
              },
              "id": "120",
              "level": 3,
              "name": "Reality monitoring",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 4,
              },
              "id": "121",
              "level": 3,
              "name": "Eyewitness testimony",
            },
          ],
          "counts": {
            "with": 0,
            "without": 10,
          },
          "id": "119",
          "level": 2,
          "name": "Real versus imagined memories",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 8,
              },
              "id": "123",
              "level": 3,
              "name": "False memories:",
            },
          ],
          "counts": {
            "with": 0,
            "without": 11,
          },
          "id": "122",
          "level": 2,
          "name": "The misinformation effect",
        },
      ],
      "counts": {
        "with": 4,
        "without": 58,
      },
      "id": "111",
      "level": 1,
      "name": "Chapter 4: Inaccuracies in memory",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "125",
          "level": 2,
          "name": "Memory and the brain",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 10,
          },
          "id": "126",
          "level": 2,
          "name": "Loss of memory after brain injury – the ‘amnesic syndrome’",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 5,
          },
          "id": "127",
          "level": 2,
          "name": "Testing amnesia",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "128",
          "level": 2,
          "name": "Psychogenic amnesia",
        },
      ],
      "counts": {
        "with": 0,
        "without": 18,
      },
      "id": "124",
      "level": 1,
      "name": "Chapter 5: Memory impairment",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "131",
          "level": 2,
          "name": "Memory development",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 9,
          },
          "id": "132",
          "level": 2,
          "name": "Memory and ageing",
        },
      ],
      "counts": {
        "with": 0,
        "without": 11,
      },
      "id": "129",
      "level": 1,
      "name": "Chapter 6: The seven ages of man",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 1,
              },
              "id": "135",
              "level": 3,
              "name": "The ‘hardware’",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 0,
              },
              "id": "136",
              "level": 3,
              "name": "The ‘software",
            },
          ],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "134",
          "level": 2,
          "name": "Can you improve your memory?",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "137",
          "level": 2,
          "name": "Rehearsal",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 0,
              },
              "id": "139",
              "level": 3,
              "name": "The benefits of spaced study",
            },
          ],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "138",
          "level": 2,
          "name": "Expanding retrieval practice",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "140",
          "level": 2,
          "name": "Meaning and memory",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 3,
              },
              "id": "142",
              "level": 3,
              "name": "Mnemonics",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "143",
              "level": 3,
              "name": "Verbal mnemonics",
            },
          ],
          "counts": {
            "with": 0,
            "without": 5,
          },
          "id": "141",
          "level": 2,
          "name": "External aids",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "144",
          "level": 2,
          "name": "Remembering names",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "145",
          "level": 2,
          "name": "Reflecting on our own learning",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "146",
          "level": 2,
          "name": "The man with a perfect memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "147",
          "level": 2,
          "name": "Final thoughts",
        },
      ],
      "counts": {
        "with": 0,
        "without": 8,
      },
      "id": "133",
      "level": 1,
      "name": "Chapter 7: Improving memory",
    },
  ],
  "counts": {
    "flashcards": {
      "learning": 14,
      "mature": 21,
      "new": 0,
    },
  },
  "id": "83",
  "name": "Memory - A Very Short Introduction",
}
`;

exports[`getSectionTreeForBook retrieves a flashcard successfully 1`] = `
{
  "children": [
    {
      "children": [],
      "counts": {
        "with": 2,
        "without": 7,
      },
      "id": "93",
      "level": 1,
      "name": "Chapter 1: You are your memory",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 1,
          },
          "id": "95",
          "level": 2,
          "name": "The logic of memory: encoding, storage, and retrieval",
        },
        {
          "children": [],
          "counts": {
            "with": 7,
            "without": 23,
          },
          "id": "96",
          "level": 2,
          "name": "Different kinds of memory: the functional structure of remembering",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 7,
          },
          "id": "97",
          "level": 2,
          "name": "Semantic, episodic, and procedural memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "98",
          "level": 2,
          "name": "Explicit and implicit memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 4,
          },
          "id": "99",
          "level": 2,
          "name": "Different kinds of memory task:",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "100",
          "level": 2,
          "name": "The experience of memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "101",
          "level": 2,
          "name": "Levels of processing",
        },
      ],
      "counts": {
        "with": 9,
        "without": 41,
      },
      "id": "94",
      "level": 1,
      "name": "Chapter 2: Mapping memories",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 0,
          },
          "id": "103",
          "level": 2,
          "name": "Inferring memory from behaviour",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "105",
              "level": 3,
              "name": "Context retrieval",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "106",
              "level": 3,
              "name": "Familiarity",
            },
          ],
          "counts": {
            "with": 3,
            "without": 6,
          },
          "id": "104",
          "level": 2,
          "name": "Retrieval: recall versus recognition",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 4,
          },
          "id": "107",
          "level": 2,
          "name": "The effect of context on recall and recognition",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "108",
          "level": 2,
          "name": "Unconscious influences on memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 6,
          },
          "id": "109",
          "level": 2,
          "name": "Categories versus continuum?",
        },
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 1,
          },
          "id": "110",
          "level": 2,
          "name": "Relating study and test",
        },
      ],
      "counts": {
        "with": 6,
        "without": 19,
      },
      "id": "102",
      "level": 1,
      "name": "Chapter 3: Pulling the rabbit out of the hat",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 3,
            "without": 3,
          },
          "id": "112",
          "level": 2,
          "name": "Forgetting",
        },
        {
          "children": [],
          "counts": {
            "with": 1,
            "without": 4,
          },
          "id": "113",
          "level": 2,
          "name": "Flashbulb memories and the reminiscence bump",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 4,
          },
          "id": "114",
          "level": 2,
          "name": "Organization and errors in memory",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 15,
              },
              "id": "116",
              "level": 3,
              "name": "Schemas – what we already know:",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "117",
              "level": 3,
              "name": "How does knowledge promote remembering?",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 4,
              },
              "id": "118",
              "level": 3,
              "name": "How can knowledge lead to errors?",
            },
          ],
          "counts": {
            "with": 0,
            "without": 21,
          },
          "id": "115",
          "level": 2,
          "name": "The effects of previous knowledge",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 5,
              },
              "id": "120",
              "level": 3,
              "name": "Reality monitoring",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 4,
              },
              "id": "121",
              "level": 3,
              "name": "Eyewitness testimony",
            },
          ],
          "counts": {
            "with": 0,
            "without": 10,
          },
          "id": "119",
          "level": 2,
          "name": "Real versus imagined memories",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 8,
              },
              "id": "123",
              "level": 3,
              "name": "False memories:",
            },
          ],
          "counts": {
            "with": 0,
            "without": 11,
          },
          "id": "122",
          "level": 2,
          "name": "The misinformation effect",
        },
      ],
      "counts": {
        "with": 4,
        "without": 58,
      },
      "id": "111",
      "level": 1,
      "name": "Chapter 4: Inaccuracies in memory",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 2,
          },
          "id": "125",
          "level": 2,
          "name": "Memory and the brain",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 10,
          },
          "id": "126",
          "level": 2,
          "name": "Loss of memory after brain injury – the ‘amnesic syndrome’",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 5,
          },
          "id": "127",
          "level": 2,
          "name": "Testing amnesia",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "128",
          "level": 2,
          "name": "Psychogenic amnesia",
        },
      ],
      "counts": {
        "with": 0,
        "without": 18,
      },
      "id": "124",
      "level": 1,
      "name": "Chapter 5: Memory impairment",
    },
    {
      "children": [
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "131",
          "level": 2,
          "name": "Memory development",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 9,
          },
          "id": "132",
          "level": 2,
          "name": "Memory and ageing",
        },
      ],
      "counts": {
        "with": 0,
        "without": 11,
      },
      "id": "129",
      "level": 1,
      "name": "Chapter 6: The seven ages of man",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 1,
              },
              "id": "135",
              "level": 3,
              "name": "The ‘hardware’",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 0,
              },
              "id": "136",
              "level": 3,
              "name": "The ‘software",
            },
          ],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "134",
          "level": 2,
          "name": "Can you improve your memory?",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "137",
          "level": 2,
          "name": "Rehearsal",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 0,
              },
              "id": "139",
              "level": 3,
              "name": "The benefits of spaced study",
            },
          ],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "138",
          "level": 2,
          "name": "Expanding retrieval practice",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "140",
          "level": 2,
          "name": "Meaning and memory",
        },
        {
          "children": [
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 3,
              },
              "id": "142",
              "level": 3,
              "name": "Mnemonics",
            },
            {
              "children": [],
              "counts": {
                "with": 0,
                "without": 2,
              },
              "id": "143",
              "level": 3,
              "name": "Verbal mnemonics",
            },
          ],
          "counts": {
            "with": 0,
            "without": 5,
          },
          "id": "141",
          "level": 2,
          "name": "External aids",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "144",
          "level": 2,
          "name": "Remembering names",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "145",
          "level": 2,
          "name": "Reflecting on our own learning",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 1,
          },
          "id": "146",
          "level": 2,
          "name": "The man with a perfect memory",
        },
        {
          "children": [],
          "counts": {
            "with": 0,
            "without": 0,
          },
          "id": "147",
          "level": 2,
          "name": "Final thoughts",
        },
      ],
      "counts": {
        "with": 0,
        "without": 8,
      },
      "id": "133",
      "level": 1,
      "name": "Chapter 7: Improving memory",
    },
  ],
  "counts": {
    "flashcards": {
      "learning": 14,
      "mature": 21,
      "new": 0,
    },
  },
  "id": "83",
  "name": "Memory - A Very Short Introduction",
}
`;
